[로또 - 1단계] 로또 자동 구매

Starlight258 opened the pull request:
안녕하세요 진영님 !
로또 1단계 미션을 완료했습니다 😁

## 궁금한 점
- 댓글에 남겼습니다!

✅ 구현된 코드에 대한 설명은 `README.md`에서 보실 수 있습니다 ! 

리뷰 부탁드립니다 !! 감사합니다~ 🙇‍♀️

Starlight258 committed:
feat: Lotto 도메인 구현

- Lotto: 1부터 45 중 무작위 6개 숫자를 포함하는 리스트를 가지는 일급 컬렉션

Starlight258 committed:
feat: Price VO 구현

- 구매 가격에 대한 값 객체(VO)를 생성하여 가격 관련 로직들을 캡슐화한다.

Starlight258 committed:
feat: InputView, OutputView 구현

Starlight258 committed:
feat: Controller, Application 구현

Starlight258 committed:
feat: Lottos 생성 및 Lotto 출력 로직 캡슐화

- List<Lotto>를 불변으로 만들어 보호하기 위해 Lottos 일급 컬렉션을 생성한다.
- outputView에서 Lottos 정보를 출력하는 두 메서드를 하나로 합친다.

Starlight258 committed:
refactor: Lotto 리팩토링

- 매직넘버 제거
- toList -> getLotto로 변경하여 뜻을 명확하게 한다.
(toList의 경우 Lotto를 List로 만드는 것인지, Lotto 내부 리스트를 가리키고 있는지 명확하지 않음)

Starlight258 committed:
refactor: 예외 메세지 명확하게 수정

- 예외 메세지에는 기대하는 결과를 적는게 더 명확한 것 같다.

Starlight258 committed:
fix: 로또 생성시 45가 포함되지 않는 문제 수정

Starlight258 committed:
test: Lotto 생성 테스트 작성

Starlight258 committed:
test: Price 테스트 작성

Starlight258 committed:
refactor: assertAll 사용하여 모든 검증 수행

Starlight258 committed:
rename: lotto 패키지 추가

- Lotto와 Lottos를 lotto 패키지로 묶는다.

Starlight258 committed:
docs: README 작성

Starlight258 committed:
refactor: Lottos 리팩토링

- toList -> getLottos로 메서드명을 변경하여 작업이 명확하게 드러나도록 한다.

Starlight258 commented on src/test/java/org/duckstudy/model/PriceTest.java:
```diff
@@ -0,0 +1,46 @@
+    @DisplayName("가격 검증 테스트")
+    class PriceValidationTest {
+
+        @Test
+        @DisplayName("가격이 0 이상일 경우 성공한다")
+        void validateSuccessWhenPriceIsEqualOrGreaterThanZero() {
+            assertThatCode(() -> new Price(1))
+                    .doesNotThrowAnyException();
+        }
```
1. 테스트 코드 작성시에 숫자가 코드에 포함되는데, 이러한 숫자들도 변수로 뜻을 표현하는 것이 좋을지 궁금합니다!
2. `@DisplayName`에서도 테스트 명세에 숫자가 포함되는데, 요구사항이 바뀌면 관리하기 어려울 것 같아 변수나 상수로 선언해서 포함시키는게 좋을지 궁금합니다!

jinyoungchoi95 replied:
> 테스트 코드 작성시에 숫자가 코드에 포함되는데, 이러한 숫자들도 변수로 뜻을 표현하는 것이 좋을지 궁금합니다!

값이 복잡하면 변수로 빼도 될 것 같은데 현재와 같이 테스트가 간단하고, 변수로 분리안해도 명확한 경우에는 굳이 안빼도 될 것 같아요.

> `@DisplayName`에서도 테스트 명세에 숫자가 포함되는데, 요구사항이 바뀌면 관리하기 어려울 것 같아 변수나 상수로 선언해서 포함시키는게 좋을지 궁금합니다!

요구사항이 변경되었을 때 `@DisplayName`이 변경되는 것은 당연해요. 테스트는 현재의 요구사항에 대한 정보들을 담고있고, 요구사항이 변경되었을 때 테스트가 깨지거나 테스트 메소드가 변경되는 것은 변경에 대해서 수정을 해야한다는 테스트의 역할을 지킨 것이니까요.

Starlight258 replied:
그렇군요 !! 테스트는 항상 현재의 요구사항을 반영해야하므로 코드 변경시 명세도 함께 수정하는 것이 자연스럽겠네요!
답변 감사합니다~

Starlight258 committed:
fix: 더 큰 범위의 예외로 수정

- 가격이 숫자로 들어오지 않을 경우와 0 미만일 경우 다시 입력받을 수 있도록 더 큰 범위의 예외로 수정

jinyoungchoi95 commented on src/main/java/org/duckstudy/model/Price.java:
```diff
@@ -0,0 +1,21 @@
+
+public record Price(int value) {
+
+    public static final int INCLUSIVE_MIN_PRICE = 0;
+    public static final int PER_PRICE = 1000;
```
```suggestion
    private static final int INCLUSIVE_MIN_PRICE = 0;
    private static final int PER_PRICE = 1000;
```
상수는 private이어야겠네요.

Starlight258 replied:
앗 맞습니다, 클래스 내부에서만 사용하므로 private으로 접근 제어자를 설정해야겠네요 !!

jinyoungchoi95 commented on src/test/java/org/duckstudy/model/lotto/LottoTest.java:
```diff
@@ -0,0 +1,24 @@
+        assertAll(
+                () -> assertThat(lotto.getLotto()).hasSize(Lotto.LOTTO_SIZE),
+                () -> assertThat(lotto.getLotto()).allMatch(
+                        number -> number >= Lotto.START_INCLUSIVE_NUMBER && number < Lotto.END_EXCLUSIVE_NUMBER)
```
위에서 이야기한 것에서 이어서 테스트의 검증값은 어딘가에 종속적이지 않아야해요. 테스트는 테스트대로 독립적이어야 테스트실행에 대한 안정성을 보장할 수 있거든요.

Starlight258 replied:
그렇군요..! 현재 테스트 코드는 검증이 상수값에 의존하게 되어있네요 ㅠ 
테스트 코드에서는 중복된 코드를 줄이고 상수로 설정해두는 것이 항상 좋은 코드는 아닌 것 같아요..😮 

아주 기본적인 코드 외에는 테스트 메서드 안에 상황 구성이나 검증값을 포함시키도록 하겠습니다!
테스트 메서드 자체가 하나의 프로그램이 되도록 작성할게요~! 감사합니다~ !!

jinyoungchoi95 commented on src/main/java/org/duckstudy/model/lotto/Lotto.java:
```diff
@@ -0,0 +1,42 @@
+        this.lotto = Collections.unmodifiableList(makeLotto());
+    }
+
+    private static List<Integer> makeNumbers() {
+        return IntStream.range(START_INCLUSIVE_NUMBER, END_EXCLUSIVE_NUMBER)
+                .boxed()
+                .toList();
+    }
```
NUMBERS 초기화의 용도로 사용되고 있는데 메소드로 제공하는 것은 사용되지 않을까의 의미로 혼동을 줄 것 같아요. static block을 사용해도 되지 않을까요?

Starlight258 replied:
그렇겠네요! 상수 리스트를 초기화하는데에만 사용한다면 static 블록을 사용하는 것이 더 명확할 것 같아요!

jinyoungchoi95 commented on src/main/java/org/duckstudy/model/lotto/Lotto.java:
```diff
@@ -0,0 +1,42 @@
+
+    public final List<Integer> lotto;
+
+    public Lotto() {
+        this.lotto = Collections.unmodifiableList(makeLotto());
+    }
```
랜덤로또만 제공하고 있으면 자동차때처럼 원하는 값을 가진 객체로 생성하기 힘든 것 아닐까요?

Starlight258 replied:
맞습니다 ! 입력받은 값으로도 로또를 생성할 수 있도록 수정하겠습니다~!

jinyoungchoi95 commented on src/main/java/org/duckstudy/model/lotto/Lotto.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    private List<Integer> makeLotto() {
+        List<Integer> numbers = new ArrayList<>(NUMBERS);
+
+        Collections.shuffle(numbers);
+        numbers = numbers.subList(FROM_INDEX, LOTTO_SIZE);
+
+        Collections.sort(numbers);
+
+        return numbers;
```
```suggestion
        Collections.shuffle(NUMBERS);
        return NUMBERS.stream()
                .limit(6)
                .sorted()
                .collect(Collectors.toList());
```
stream을 활용한다면 이렇게 표현될 수 있을 것 같아요. 개인적으로 메소드 내에서 `new ArrayList<>()`를 사용할 수 밖에 없는 로직의 경우 어쩔 수 없지만 가변적인 list를 내부에서 컨트롤했을 때 값 자체가 변경될 확률이 너무 높아서 지양하는 편이에요.

Starlight258 replied:
그렇군요 !! 불변 리스트로 가변 리스트를 만들더라도 가변 리스트는 언제나 변경이 가능하기 때문에 위험하네요 !
`stream`은 연산시에 원본을 변경하지 않으니 불변 리스트에 stream을 사용하는 것이 좋겠네요!
`불변 리스트 + stream` 조합 기억하겠습니다,, 감사합니다~!!

jinyoungchoi95 commented on src/main/java/org/duckstudy/model/lotto/Lottos.java:
```diff
@@ -0,0 +1,17 @@
+import java.util.Collections;
+import java.util.List;
+
+public class Lottos {
```
일급컬랙션이 지금 스펙에서 당장 필요한 객체인지는 의문이네요.

Starlight258 replied:
뽑은 로또들에 대한 처리가 다음 미션에 있을 것 같아서 미리 만들어두었는데 당장은 필요 없을 것 같긴 합니다..! ㅠㅠ

jinyoungchoi95 reviewed (changes requested):
안녕하세요 명지님 😃

몇가지 코멘트남겨두었으니 확인부탁드려요.

감사합니다🙇‍♂️

Starlight258 committed:
refactor: 상수 접근제어자 private으로 변경

- 클래스 내부에서만 사용하므로 상수는 private 접근제어자로 둔다.
- 만약 public 상수를 선언할 경우에는 enum으로 정의하는 것이 타입 안정성을 제공하여 좋다.

Starlight258 committed:
test: 검증값을 상수값이 아닌 값으로 표현하도록 수정

- 상수나 변수로 검증값을 둘 경우 테스트시 필요없는 의존성이 생겨 지양한다.
- 검증은 최대한 간단하게 수행한다. 변수나 상수를 이용하여 값 검증시 해당 값을 확인하기 위해 코드 위쪽까지 거슬러 올라가야한다.

Starlight258 committed:
refactor: Lotto 리팩토링

- 상수 리스트 생성시에만 사용되는 작업은 static 블록으로 이동한다.
- 수동 로또 생성을 위한 정적 팩토리 메서드를 제공한다.
- 불변 리스트 + stream 조합을 사용하여 불변성을 지키며 결과 리스트를 얻는다.
- Collections.shuffle은 원본 리스트를 변경하므로 stream 연산에서 무작위 정수를 사용해 로또를 얻도록 수정한다.

Starlight258 committed:
refactor: stream을 이용한 Lotto 리스트 생성

- Stream을 이용해 Lotto 리스트를 생성하고 toList()를 호출하여 불변 리스트를 반환한다.
- 당장 사용하지 않는 Lottos는 제거한다.

Starlight258 commented:
@jinyoungchoi95
리뷰해주신 사항들 덕분에 많이 배웠습니다 🙇‍♀️
감사합니다 ! 😁

jinyoungchoi95 commented on src/main/java/org/duckstudy/model/lotto/Lotto.java:
```diff
@@ -0,0 +1,45 @@
+        return new Lotto(result);
+    }
+
+    public static Lotto createManualLotto(List<Integer> lotto) {
+        return new Lotto(lotto);
+    }
```
`manualLotto` 네이밍에서 의미가 담겨있기는 하지만 개인적으로 생성자와 동일한 성격의 정적 팩토리 메소드는 생성하지 않아도 된다고 생각해요. 이미 생성자가 있는데 굳이의 느낌이 있기 때문이에요.

보통 이펙티브 자바에서도 정적 팩토리 메소드에 대한 이야기를 하는데, 저는 생성자에서 만들 수 있으면 가능하면 생성자를 사용하고 `createRandomLotto`와 같이 정말 생성말고도 추가 로직이 필요한 경우 의미전달이 필요할때만 정적팩토리가 사용되어야한다고 생각해요.

Starlight258 replied:
그렇군요 !! 단순 객체를 생성하는데 정적 팩토리 메서드를 사용하면서 코드가 불필요하게 길어진 적이 있어서 
공감합니다!! 생성자와 차별화되는 이점이 있을 때 정적 팩토리 메서드를 사용하겠습니다 감사해요 🙇‍♀️

jinyoungchoi95 reviewed (approved):
안녕하세요 명지님 😃

피드백 잘 반영해주셨네요. 다음 미션 진행을 위해 머지하겠습니다.

감사합니다🙇‍♂️

Starlight258 committed:
refactor: 불필요한 정적 팩토리 메서드 제거

- 단순 객체 생성은 생성자로 충분하기 때문에 정적 팩토리 메서드를 제거한다.
